internal/blobstore/blobstore.go

```diff
@@ -76,22 +76,65 @@ func NewContentChallengeResponse(chal *ContentChallenge, r io.ReadSeeker) (*Cont
 	}, nil
 }
 
+// Store stores data blobs in the configured store.
+type Store struct {
+	store
+}
+
 // Store stores data blobs in mongodb, de-duplicating by
 // blob hash.
-type Store struct {
+type store interface {
+	Put(r io.Reader, name string, size int64, hash string, proof *ContentChallengeResponse) (*ContentChallenge, error)
+	PutUnchallenged(r io.Reader, name string, size int64, hash string) error
+	Open(name string) (ReadSeekCloser, int64, error)
+	Remove(name string) error
+}
+
+// gridStore stores data blobs in mongodb gridfs, de-duplicating by
+// blob hash.
+type gridStore struct {
 	mstore blobstore.ManagedStorage
+	db     *mgo.Database
+	prefix string
 }
 
 // New returns a new blob store that writes to the given database,
 // prefixing its collections with the given prefix.
+// This is the legacy function. Prefer using NewGridFSFromProviderConfig
 func New(db *mgo.Database, prefix string) *Store {
 	rs := blobstore.NewGridFS(db.Name, prefix, db.Session)
 	return &Store{
-		mstore: blobstore.NewManagedStorage(db, rs),
+		&gridStore{
+			mstore: blobstore.NewManagedStorage(db, rs),
+			db:     db,
+			prefix: prefix,
+		},
+	}
+}
+
+// NewBlobstoreFromProviderConfig returns a new blob stor that writes to mongodb
+// as defined in the given ProviderConfig
+func NewBlobstoreFromProviderConfig(pc *ProviderConfig) *Store {
```

Comment:
I call it Blobstore, since it is not directly GridFS, it is the github.com/juju/blobstore layer that is built upon GridFS. I'd rather use a name which reflects that it is not direct GridFS use.